### PR TITLE
qemu_guest_agent: new case to get virtio device of vm

### PR DESCRIPTION
**Pull request by xiagao**
Get virtio device of windows guest.
ID: 1824430
Signed-off-by: Xiaoling Gao <xiagao@redhat.com>

### CODE REVIEW

**Comment by xiagao** [comment]
depends on: https://github.com/avocado-framework/avocado-vt/pull/2524

**Comment by xiagao** [comment]
@lolyu  Call for review, thanks..

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -2686,6 +2686,63 @@ def bl_check(qga_cmd):
...
+            if rsult_qga != rsult_guest:
+                msg = "The result is different between qga and guest\n"
+                msg += "from qga: %s\n" % rsult_qga
+                msg += "from guest: %s\n" % rsult_guest
+                test.fail(msg)
+
+        devs_list = self.gagent.get_virtio_device()
+        for device in devs_list:
+            driver_name = device["driver-name"]
+            error_context.context("Check %s info." % driver_name, logging.info)
```
Comment (lolyu) [inline qemu/tests/qemu_guest_agent.py:2711]: here better use `logging.info`

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -2686,6 +2686,63 @@ def bl_check(qga_cmd):
...
+
+            driver_date = device["driver-date"]
+            driver_version = device["driver-version"]
+            device_address = device["address"]["data"]
+            device_id = device_address["device-id"]
+            vendor_id = device_address["vendor-id"]
+            check_driver_cmd = params["check_driver_powershell_cmd"]
```
Comment (lolyu) [inline qemu/tests/qemu_guest_agent.py:2718]: better put LINE 2718 out of this `for` loop.

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -2686,6 +2686,63 @@ def bl_check(qga_cmd):
...
+                                                       driver_name)
+            else:
+                check_driver_cmd = check_driver_cmd % ("devicename",
+                                                       driver_name)
+            driver_info_guest = session.cmd_output(check_driver_cmd)
+            # check driver date
+            date = re.search(r"driverdate.*\:\s(\S{8})", driver_info_guest,
+                             re.I).group(1)
+            date_l = [date[0:4], date[4:6], date[6:8]]
+            driver_date_guest = "-".join(date_l)
```
Comment (lolyu) [inline qemu/tests/qemu_guest_agent.py:2730]:
```python
date = re.search(r"driverdate.*\:\s(\d{4})(\d{2})(\d{2})")
driver_date_guest = "-".join(data.groups())
```

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -2686,6 +2686,63 @@ def bl_check(qga_cmd):
...
+            _result_check(driver_date, driver_date_guest)
+
+            # check driver version
+            driver_ver_guest = re.search(r"driverversion.*\:\s(\S+)",
```
Comment (lolyu) [inline qemu/tests/qemu_guest_agent.py:2734]: if the version is a digit, better to use `\d` instead of `\S`

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -2686,6 +2686,63 @@ def bl_check(qga_cmd):
...
+                                         driver_info_guest, re.I).group(1)
+            _result_check(driver_version, driver_ver_guest)
+
+            # check vender id and device id
+            id = re.search(r"deviceid.*VEN_([A-Za-z0-9]+)&DEV_([A-Za-z0-9]+)&",
```
Comment (lolyu) [inline qemu/tests/qemu_guest_agent.py:2739]: `id` is a builtin function, better to use other names like `id_result`

File: qemu/tests/qemu_guest_agent.py (hunk 2 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: update

File: qemu/tests/qemu_guest_agent.py (hunk 3 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: update

File: qemu/tests/qemu_guest_agent.py (hunk 4 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py:2733]: driverversion : 100.82.104.18000

File: qemu/tests/qemu_guest_agent.py (hunk 5 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: ok

**Comment by xiagao** [comment]
@lolyu  updated already.

**Comment by xiagao** [comment]
@vivianQizhu Call for your review..

File: qemu/tests/qemu_guest_agent.py [hunk 6]
```diff
@@ -2686,6 +2686,62 @@ def bl_check(qga_cmd):
...
+            driver_version = device["driver-version"]
+            device_address = device["address"]["data"]
+            device_id = device_address["device-id"]
+            vendor_id = device_address["vendor-id"]
+            if "Ethernet" in driver_name:
+                check_driver_cmd = check_driver_cmd_org % ("friendlyname",
+                                                           driver_name)
+            else:
+                check_driver_cmd = check_driver_cmd_org % ("devicename",
+                                                           driver_name)
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:2724]: What about:
```
filter_name = "friendlyname" if "Ethernet" in driver_name else "devicename"
check_driver_cmd = check_driver_cmd_org % (filter_name, driver_name)
```

File: qemu/tests/qemu_guest_agent.py [hunk 7]
```diff
@@ -2686,6 +2686,62 @@ def bl_check(qga_cmd):
...
+            driver_info_guest = session.cmd_output(check_driver_cmd)
+            # check driver date
+            date_group = re.search(r"driverdate.*\:\s(\d{4})(\d{2})(\d{2})",
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:2727]: Could you please provide an example of this format? I am not running it to check by myself, thanks.

File: qemu/tests/qemu_guest_agent.py [hunk 8]
```diff
@@ -2686,6 +2686,62 @@ def bl_check(qga_cmd):
...
+            # check driver version
+            driver_ver_guest = re.search(r"driverversion.*\:\s(\S+)",
+                                         driver_info_guest, re.I).group(1)
+            _result_check(driver_version, driver_ver_guest)
+
+            # check vender id and device id
+            pattern_dev = r"deviceid.*VEN_([A-Za-z0-9]+)&DEV_([A-Za-z0-9]+)&"
+            id_dev = re.search(pattern_dev, driver_info_guest, re.I)
+            vender_id_guest = int(id_dev.group(1), 16)
+            device_id_guest = int(id_dev.group(2), 16)
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:2741]: Is this format all the same for different version windows?

File: qemu/tests/qemu_guest_agent.py [hunk 9]
```diff
@@ -2686,6 +2686,62 @@ def bl_check(qga_cmd):
...
+
+        def _result_check(rsult_qga, rsult_guest):
+            if rsult_qga != rsult_guest:
+                msg = "The result is different between qga and guest\n"
+                msg += "from qga: %s\n" % rsult_qga
+                msg += "from guest: %s\n" % rsult_guest
+                test.fail(msg)
+
+        devs_list = self.gagent.get_virtio_device()
+        check_driver_cmd_org = params["check_driver_powershell_cmd"]
```
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py:2709]: I mean original, in Line2720 and L2723 ，the cmd will be changed.

File: qemu/tests/qemu_guest_agent.py (hunk 6 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py:2724]: yes, it's better,will update.

File: qemu/tests/qemu_guest_agent.py (hunk 7 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py:2727]: the format is: 
driverdate    : 20200219000000.******+***

File: qemu/tests/qemu_guest_agent.py (hunk 8 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py:2741]: yes, virtio device has this format in different windows os.

File: qemu/tests/qemu_guest_agent.py [hunk 10]
```diff
@@ -2686,6 +2686,61 @@ def bl_check(qga_cmd):
...
+            driver_name = device["driver-name"]
+            error_context.context("Check %s info." % driver_name, logging.info)
+
+            driver_date = device["driver-date"]
+            driver_version = device["driver-version"]
+            device_address = device["address"]["data"]
+            device_id = device_address["device-id"]
+            vendor_id = device_address["vendor-id"]
+
+            filter_name = "friendlyname" if "Ethernet" in driver_name else "devicename"
```
Comment (vivianQizhu) [inline qemu/tests/qemu_guest_agent.py:2720]: I am afraid this line is too long.

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: yes, modified..thx.

**Comment by xiagao** [comment]
@vivianQizhu 

 (1/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_virtio_device.default: PASS (255.84 s)
 (2/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_virtio_device.with_input_device: PASS (190.04 s)
 (3/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_virtio_device.default: PASS (185.03 s)
 (4/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_virtio_device.with_input_device: PASS (122.66 s)
 (5/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.Win8.i386.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_virtio_device.default: PASS (92.23 s)
 (6/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_blk.up.virtio_net.Guest.Win8.i386.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_virtio_device.with_input_device: PASS (88.52 s)

**Pull request merged**